bucketeer/src/main/kotlin/io/bucketeer/sdk/android/internal/event/EventCreators.kt

```diff
@@ -103,94 +104,120 @@ internal fun newGoalEvent(
   )
 }
 
-internal fun newGetEvaluationLatencyMetricsEvent(
+internal fun newMetricsEvent(
   clock: Clock,
   idGenerator: IdGenerator,
-  seconds: Long,
-  featureTag: String,
+  featureTag: String?,
   appVersion: String,
+  metricsEventType: MetricsEventType,
+  apiID: ApiID,
+  // note: only available on success request
+  latencySecond: Long? = null,
+  // note: only available on success request
+  sizeByte: Int? = null,
 ): Event {
   return Event(
     id = idGenerator.newId(),
     type = EventType.METRICS,
     event = EventData.MetricsEvent(
       timestamp = clock.currentTimeSeconds(),
-      type = MetricsEventType.GET_EVALUATION_LATENCY,
-      event = MetricsEventData.GetEvaluationLatencyMetricsEvent(
-        labels = mapOf(
-          "tag" to featureTag,
-        ),
-        duration = seconds,
-      ),
+      type = metricsEventType,
+      event = newMetricsEventData(featureTag, apiID, metricsEventType, latencySecond, sizeByte),
       sdkVersion = BuildConfig.SDK_VERSION,
       metadata = newMetadata(appVersion),
     ),
   )
 }
+internal fun newMetricsEventData(
+  featureTag: String?,
+  apiID: ApiID,
+  type: MetricsEventType,
+  // note: only available on success request
+  latencySecond: Long? = null,
+  // note: only available on success request
+  sizeByte: Int? = null,
+): MetricsEventData {
+  // note: featureTag only available from `GET_EVALUATIONS`
+  val labels = if (featureTag != null) mapOf("tag" to featureTag) else mapOf()
```

Comment:
Hi @cre8ivejp 
I was thinking about  _**"featureTag" could be `null` and the `REGISTER_EVENTS` API did not have the tag**_

1. **We register events to a server in batch**. 
That means we may have many "featureTag" for each event in the batch.
Min 1, max = batch_size ( the default is 4).
Did we support `tag` could be an array of String?  I don't think we will create multiple error event metrics for each "featureTag"  for a single API calling.

2. I was checking on the iOS repo and JS repo for finding a reference 
- JS Repo didn't track success | error metric on [**the "register_event" call**](https://github.com/bucketeer-io/javascript-client-sdk/pull/17/files#diff-72ac04774c2e42de828a1d38c8db97d943013e1c41d21aa130eb0d6078c66182) . So it could not use to make decisions and references.
- iOS Repo tracks the `REGISTER_EVENTS` API call, but it didn't set a tag when tracking the success | error metric. [EventInteractor.swift:line169](https://github.com/bucketeer-io/ios-client-sdk/blob/af5709fd5fb171c605fe294bfecf9b26b512c4d2/Bucketeer/Sources/Internal/Event/EventInteractor.swift#L169)

What do you think ?

Reply:
https://github.com/bucketeer-io/android-client-sdk/blob/25b3ed6646c58618dc8c2c10acf9577847b27748/bucketeer/src/main/kotlin/io/bucketeer/sdk/android/BKTConfig.kt#L18

`BKTConfig.featureTag` is non-null, and we throw error when it is not configured.

https://github.com/bucketeer-io/android-client-sdk/blob/25b3ed6646c58618dc8c2c10acf9577847b27748/bucketeer/src/main/kotlin/io/bucketeer/sdk/android/BKTConfig.kt#L90

So `featureTag` should always exist.

Reply:
> Could `featureTag` have many different values?
> Could the SDK's user change the `featureTag`?

`BKTConfig` is a data class, and `featureTag` is declared as `val`, so it's immutable.

`BKTClient` should be recreated when changing the `featureTag`.

Reply:
that means we may have a case that will has many `featureTag` values  @yshrsmz

Reply:
> > On the BKTConfig.kt, featureTag could be null
> 
> https://github.com/bucketeer-io/android-client-sdk/blob/25b3ed6646c58618dc8c2c10acf9577847b27748/bucketeer/src/main/kotlin/io/bucketeer/sdk/android/BKTConfig.kt#L18
> 
> `BKTConfig.featureTag` is non-null, and we throw error when it is not configured.
> 
> https://github.com/bucketeer-io/android-client-sdk/blob/25b3ed6646c58618dc8c2c10acf9577847b27748/bucketeer/src/main/kotlin/io/bucketeer/sdk/android/BKTConfig.kt#L90
> 
> So `featureTag` should always exist.

thank you, I was looking in the wrong code place. (The builder class)

Reply:
> That means we may have a case that will has many featureTag values @yshrsmz

Yes, but that shouldn't be a problem as we cache `featureTag` in each event.

Reply:
> > That means we may have a case that will has many featureTag values @yshrsmz
> 
> Yes, but that shouldn't be a problem as we cache `featureTag` in each event.

Yes, I know.
But it will lead to how to deal with this case. 
When I checked iOS & JS SDK repo. I thought we dismissed the tag when calling `register_events` in batch. Because all of them merged in the codebase. So it may be a new problem for iOS and JS SDK to solve.

### Solution 1

- I will collect all `featureTag` from the batch call.
- For each `featureTag` will create the correct error event metric. In most cases, there will be only one `featureTag`

What do you think? @yshrsmz  @cre8ivejp

Reply:
I'm not sure if I understand correctly, but each event should have a `featureTag`(from BKTConfig) of the time it is recorded. You always use featureTag from BKTConfig.

So register_events request may include many different `featureTag`s, but success/failure metrics event of that register_events request does nothing to do with `featureTag`s inside the request. You can just use `BKTConfig.featureTag` to create success/failure event.

Reply:
@cre8ivejp 
on your example : 
- In the DB has metric events with the tag "android"
 `["tag: android", "tag: android"]`
**If sending them fails will create an error metric event with the tag "android"** 
- before sending them, the app initializes the SDK again using a different tag, "android-tv"
- after that, some metric events were tracked with the new tag "android-tv" 
- Now in the DB, we have metrics events with tags "android", and "android-tv".
 `["tag: android", "tag: android", "tag: android-tv", "tag: android-tv"]`
- Now it is time to **"register_events" in batch** ... and we may get some errors.
- The question: new error metric will be tracked with what tags? 
  1) Option 1:  Tags cached in the database. One error metric event with the tag "android", and one error metric event with the tag "android-tv"
  2) Option 2: Or tag in the current BKTConfig.featureTag "android-tv"
  
If 1) we will have more than 1 new error metric event tracked depending on how many unique `featured tags`  
If 2) we will only have one error metric, but the tag only has the current, not the previous cached.

The problem with (2) is, error metrics events have a tag that seems not correct with cached events when they tracked to DB

Hope it clearly my question.

Reply:
Say you instantiated BKTClient with "android-3" featureTag, and you have 2 events in DB. 
BKTClient tries to create a register_events request.

```js
{
  "events": [
    {
      "eventType": 4, // METRICS
      "event": {
        "labels": { "tag": "android" },
        // other values go here...
      }
    }, 
    {
      "eventType": 1, // GOAL
      "event": {
        "labels": { "tag": "android-2" },
        // other values go here...
      }
    }
  ]
}
```

And the response is returned with error(all registration is failed), so you save failure metrics event.

Then after several minutes, BKTClient tries to send another register_events. In this case, we have 3 events in a DB.

The request looks like this.


```js
{
  "events": [
    {
      "eventType": 4, // METRICS
      "event": {
        "labels": { "tag": "android" },
        // other values go here...
      }
    }, 
    {
      "eventType": 1, // GOAL
      "event": {
        "labels": { "tag": "android-2" },
        // other values go here...
      }
    },
    {
      "eventType": 4, // METRICS
      "event": {
        // here you save new failure metrics event with `android-3` tag
        "labels": { "tag": "android-3" },
        // other values go here...
      }
    }
  ]
}
```

So feature tag is in each event, and you don't need to care about multiple featureTag in a single `labels` map.